File: aura-impl/src/main/resources/aura/value/ExpressionFunctions.js
/*
 * Copyright (C) 2012 salesforce.com, inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/*
 * client implementations of all the expression functions
 */
var expressionFunctions = {
    eq: function equals(args) {
    	var a = args[0];
    	var b = args[1];
    	a = a?a.unwrap():a;
    	b = b?b.unwrap():b;
        if (aura.util.isUndefinedOrNull(a)) {
            return aura.util.isUndefinedOrNull(b);
        }
        return a === b;
    },

    ne: function notequals(args) {
        return !expressionFunctions.eq(args);
    },

    "if": function ternary(args) {
    	var a = args[0];
    	a = a?a.unwrap():a;
        return a ? args[1] : args[2];
    },

    // add also is string concatenate
    add: function add(args) {
    	var a = args[0];
    	var b = args[1];
    	a = a?a.unwrap():a;
    	b = b?b.unwrap():b;
        if (aura.util.isUndefinedOrNull(a)) {
            if (aura.util.isString(b)) {
                return b;
            } else if (aura.util.isUndefinedOrNull(b)) {
                return "";
            }
        } else if (aura.util.isUndefinedOrNull(b) && aura.util.isString(a)) {
            return a;
        }
        return a + b;
    },

    sub: function subtract(args) {
    	var a = args[0];
    	var b = args[1];
    	a = a?a.unwrap():a;
    	b = b?b.unwrap():b;
        return a - b;
    },

    mult: function multiply(args) {
    	var a = args[0];
    	var b = args[1];
    	a = a?a.unwrap():a;
    	b = b?b.unwrap():b;
        return a * b;
    },

    div: function divide(args) {
    	var a = args[0];
    	var b = args[1];
    	a = a?a.unwrap():a;
    	b = b?b.unwrap():b;
        return a / b;
    },

    mod: function modulus(args) {
    	var a = args[0];
    	var b = args[1];
    	a = a?a.unwrap():a;
    	b = b?b.unwrap():b;
        return a % b;
    },

    neg: function negate(args) {
    	var a = args[0];
    	a = a?a.unwrap():a;
        return -a;
    },

    abs: function absolutevalue(args) {
    	var a = args[0];
    	a = a?a.unwrap():a;
        return Math.abs(a);
    },

    gt: function greaterthan(args) {
    	var a = args[0];
    	var b = args[1];
    	a = a?a.unwrap():a;
    	b = b?b.unwrap():b;
        return a > b;
    },

    lt: function lessthan(args) {
    	var a = args[0];
    	var b = args[1];
    	a = a?a.unwrap():a;
    	b = b?b.unwrap():b;
        return a < b;
    },

    ge: function greaterthanorequalto(args) {
    	var a = args[0];
    	var b = args[1];
    	a = a?a.unwrap():a;
    	b = b?b.unwrap():b;
        return a >= b;
    },

    le: function lessthanorequalto(args) {
    	var a = args[0];
    	var b = args[1];
    	a = a?a.unwrap():a;
    	b = b?b.unwrap():b;
        return a <= b;
    },

    and: function and(args) {
    	var a = args[0];
    	var b = args[1];
    	a = a?a.unwrap():a;
    	b = b?b.unwrap():b;
        return a && b;
    },

    or: function or(args) {
    	var a = args[0];
    	var b = args[1];
    	a = a?a.unwrap():a;
    	b = b?b.unwrap():b;
    	return a || b;
    },

    not: function not(args) {                             
    	var a = args[0];
        a = a ? a.unwrap() : a;
        return a !== true;
    }
};

//#include aura.value.ExpressionFunctions_export

File: aura-impl-expression/src/main/java/org/auraframework/impl/expression/ExpressionFunctions.java
/*
 * Copyright (C) 2013 salesforce.com, inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.auraframework.impl.expression;

import static org.auraframework.impl.expression.functions.BooleanFunctions.AND;
import static org.auraframework.impl.expression.functions.BooleanFunctions.NOT;
import static org.auraframework.impl.expression.functions.BooleanFunctions.OR;
import static org.auraframework.impl.expression.functions.BooleanFunctions.TERNARY;
import static org.auraframework.impl.expression.functions.MathFunctions.ABSOLUTE;
import static org.auraframework.impl.expression.functions.MathFunctions.DIVIDE;
import static org.auraframework.impl.expression.functions.MathFunctions.MODULUS;
import static org.auraframework.impl.expression.functions.MathFunctions.MULTIPLY;
import static org.auraframework.impl.expression.functions.MathFunctions.NEGATE;
import static org.auraframework.impl.expression.functions.MathFunctions.SUBTRACT;
import static org.auraframework.impl.expression.functions.MultiFunctions.ADD;
import static org.auraframework.impl.expression.functions.MultiFunctions.EQUALS;
import static org.auraframework.impl.expression.functions.MultiFunctions.GREATER_THAN;
import static org.auraframework.impl.expression.functions.MultiFunctions.GREATER_THAN_OR_EQUAL;
import static org.auraframework.impl.expression.functions.MultiFunctions.LESS_THAN;
import static org.auraframework.impl.expression.functions.MultiFunctions.LESS_THAN_OR_EQUAL;
import static org.auraframework.impl.expression.functions.MultiFunctions.NOTEQUALS;

import java.util.Map;

import org.auraframework.annotations.Annotations.ServiceComponent;
import org.auraframework.impl.expression.functions.Empty;
import org.auraframework.impl.expression.functions.Format;
import org.auraframework.impl.expression.functions.Function;
import org.auraframework.impl.expression.functions.Join;
import org.auraframework.impl.expression.functions.Token;
import org.auraframework.service.ContextService;
import org.auraframework.service.DefinitionService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Lazy;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableMap.Builder;

/**
 * Service used to lookup {@link Function}s by name
 */
@Lazy
@ServiceComponent("expressionFunctions")
public class ExpressionFunctions {

    private final Map<String, Function> functionsByName;

    /**
     * @param contextService The service for interacting with the context
     * @param definitionService The service for looking up components.
     */
    @Lazy
    @Autowired
    public ExpressionFunctions(final ContextService contextService, final DefinitionService definitionService) {
        final Function[] l = new Function[]{
            ADD,
            SUBTRACT,
            MULTIPLY,
            DIVIDE,
            MODULUS,
            GREATER_THAN,
            GREATER_THAN_OR_EQUAL,
            LESS_THAN,
            LESS_THAN_OR_EQUAL,
            AND,
            OR,
            NOT,
            NEGATE,
            ABSOLUTE,
            EQUALS,
            NOTEQUALS,
            TERNARY,
            Empty.getInstance(),
            Format.getInstance(),
            Token.getInstance(contextService, definitionService),
            Join.getInstance()
        };
        final Builder<String, Function> b = ImmutableMap.builder();
        for (final Function f : l) {
            for (final String k : f.getKeys()) {
                b.put(k, f);
            }
        }
        functionsByName = b.build();
    }

    /**
     * @param name The name of the function to lookup.
     * @return The matching {@link Function}.
     */
    public Function lookup(final String name) {
        return functionsByName.get(name);
    }
}
